onnxruntime/core/optimizer/transformer_memcpy.cc

```diff
@@ -274,22 +288,37 @@ static const onnxruntime::NodeArg* FindNodeArg(const NodeArgSetType& def_set, co
 // We duplicate any initializer that is used by both provider nodes and non-provider nodes
 // to ensure that provider nodes and non-provider nodes don't share initializers, as they
 // need to stay in different memory locations.
-void TransformerMemcpyImpl::ProcessInitializers(const KernelRegistryManager& kernel_registries) {
+bool TransformerMemcpyImpl::ProcessInitializers(const KernelRegistryManager& kernel_registries, const InitializedTensorSet& initializers_consumed) {
   std::map<const onnxruntime::NodeArg*, onnxruntime::NodeArg*> replacements;
-  for (const auto& pair : graph_.GetAllInitializedTensors()) {
+  for (const auto& pair : initializers_consumed) {
     const auto& name = pair.first;
     const onnxruntime::NodeArg* provider_def = FindNodeArg(provider_input_defs_, name);
     const onnxruntime::NodeArg* non_provider_def = FindNodeArg(non_provider_input_defs_, name);
     if (provider_def != nullptr && non_provider_def != nullptr) {
       std::string new_def_name = graph_.GenerateNodeArgName(name);
       auto& new_def = graph_.GetOrCreateNodeArg(new_def_name, provider_def->TypeAsProto());
 
-      const TensorProto* tensor_proto = nullptr;
-      bool found = graph_.GetInitializedTensor(name, tensor_proto);
-      ORT_ENFORCE(found, "Failed to get initialized tensor ", name);
+      const TensorProto* tensor_proto = pair.second;
+      ORT_ENFORCE(tensor_proto != nullptr, "Failed to get initialized tensor ", pair.first);
 
       TensorProto new_tensor_proto = *tensor_proto;
       *(new_tensor_proto.mutable_name()) = new_def_name;
+
+      // We make a copy of the initializer that is to be consumed by the provider Node so that
+      // session state initializer can copy it over to the provider device during its operation
+      LOGS(logger_, INFO) << "Making a copy of the initializer: " << name << " and the copy's name is: " << new_def_name;
+
+      if (graph_.IsSubgraph()) {
+          // TODO: The copy being made is possibly redundant 
+          // When multiple subraphs consume the same initializer as an implicit input,
+          // mulitple copies of the initializer will be made into the provider device
```

Comment:
>mulitple [](start = 13, length = 8)

multiple
[https://ewsoftware.github.io/VSSpellChecker/html/027d2fbc-7bfb-4dc3-b4f5-85f95fcf7629.htm](https://ewsoftware.github.io/VSSpellChecker/html/027d2fbc-7bfb-4dc3-b4f5-85f95fcf7629.htm) is very helpful

Reply:
Fixed.

Reply:
Thanks for this suggestion. It is pretty helpful.